tket2/src/portmatching/pattern.rs

```diff
@@ -146,35 +62,387 @@ pub enum InvalidPattern {
+
+    for out_port in hugr.node_outputs(inp) {
+        let mut linked_ports = hugr.linked_inputs(inp, out_port);
+        let in_port = linked_ports.find_map(|(n, p)| (n == out).then_some(p));
+
+        if let Some(in_port) = in_port {
+            return Err(InvalidPattern::EmptyWire {
+                from_node: inp,
+                from_port: out_port.into(),
+                to_node: out,
+                to_port: in_port.into(),
+            });
+        }
+    }
+    Ok(())
+}
+
+/// Turn a path map for nodes into a map for every variable in `hugr`.
+///
+/// All non-IO nodes in hugr must be bound in `path_map`.
+fn into_variable_map(path_map: NodeToPathMap, circuit: &Circuit<impl HugrView>) -> VariableMap {
+    let hugr = circuit.hugr();
+    let mut ret = BTreeMap::new();
+    let all_nodes = circuit.commands().map(|cmd| cmd.node()).collect_vec();
+
+    // 1. Add nodes
+    for &node in &all_nodes {
+        let path = path_map[&node];
+        let val = HugrVariableValue::Node(node);
+        let var = HugrNodeID::new(path);
+        ret.insert(val, var.into());
+    }
+
+    // 2. Add ports
+    let in_ports = all_nodes
+        .iter()
+        .flat_map(|&n| filtered_node_inputs(n, hugr).map(move |p| (n, hugr::Port::from(p))));
+    let out_ports = all_nodes
+        .iter()
+        .flat_map(|&n| filtered_node_outputs(n, hugr).map(move |p| (n, hugr::Port::from(p))));
+    let ports = in_ports.chain(out_ports);
+    for (node, port) in ports {
+        let path = HugrNodeID::new(path_map[&node]);
+        let var = HugrPortID::new(path, port);
+        let val = (node, port);
+        ret.insert(val.into(), var.into());
+    }
+
+    ret
+}
+
+/// Fix the incoming port names (i.e. their path)
+///
+/// Choose the root that minimises the maximum path length.
+fn get_node_to_path_map(circuit: &Circuit<impl HugrView>) -> Result<NodeToPathMap, InvalidPattern> {
+    if circuit.num_operations() == 0 {
+        return Err(InvalidPattern::EmptyCircuit);
+    }
+    let nodes = circuit.commands().map(|cmd| cmd.node());
+    let mut best_map = None;
+    let max_path = |map: &BTreeMap<_, _>| map.values().max().copied().unwrap();
+    for root in nodes {
+        // Use Dijkstra to find all shortest paths from root
+        let Some(map) = dijkstra_all_shortest_paths(root, circuit) else {
+            continue;
+        };
+        if best_map.is_none() || max_path(&map) < max_path(best_map.as_ref().unwrap()) {
+            best_map = Some(map);
+        }
+    }
+    let Some(best_map) = best_map else {
+        return Err(InvalidPattern::NotConnected);
+    };
+    Ok(best_map)
 }
 
-impl From<NoRootFound> for InvalidPattern {
-    fn from(_: NoRootFound) -> Self {
-        InvalidPattern::NotConnected
+/// Compute the minimum paths from the root to all nodes.
+///
+/// The returned map is a valid [`NodeToPathMap`], i.e. it has a key for every
+/// node in `hugr`.
+fn dijkstra_all_shortest_paths(
```

Comment:
It feels like we should be able to use either portgraph or petgraph for this, but it may not be worth the hassle.

Reply:
Yes you are right -- although what makes it difficult is that we keep track of the builder object along the shortest path, so there isn't a clean way to associate each edge with a weight. Rather, each edge corresponds to an "action" on the builder obtained so far.

So as you say, I don't think it's worth the hassle. I've added a comment saying this.